Java error from production changes in concrete syntax

Adding or removing a production in a concrete grammar is giving me a number of errors like this:
```
grammars:
    [javac] Compiling 197 source files to /home/micha576/research/melt/silver/generated/bin
    [javac] Ignoring source, target and bootclasspath as release has been set
    [javac] /home/micha576/research/melt/silver/generated/src/ComScaN/host/concreteSyntax/Pp_ComScaN_host_concreteSyntax_Concrete_sv_105_0.java:22: error: cannot find symbol
    [javac]     public static final int num_local_attrs = Init.count_local__ON__ComScaN_host_concreteSyntax_p_ComScaN_host_concreteSyntax_Concrete_sv_105_0;
    [javac]                                                   ^
    [javac]   symbol:   variable count_local__ON__ComScaN_host_concreteSyntax_p_ComScaN_host_concreteSyntax_Concrete_sv_105_0
    [javac]   location: class Init
```
Removing the generated Java sources in the Silver `generated` directory fixes this, so I suspect it is not correctly detecting changes to the grammar and trying to reuse out-of-date code.

I have not seen this error without having equations for abstract syntax trees in the productions.

--- a/grammars/silver/compiler/translation/java/driver/BuildProcess.sv
+++ b/grammars/silver/compiler/translation/java/driver/BuildProcess.sv
@@ -246,7 +246,8 @@ IO<()> ::= silverGen::String keepFiles::[String] r::Decorated RootSpec
         exit(-5);
       });
     });
-    oldSrcFiles::[String] <- listContents(srcPath);
+    srcDirContents::[String] <- listContents(srcPath);
+    oldSrcFiles::[String] <- filterM(isFile, map(append(srcPath, _), srcDirContents));
     deleteFiles(removeAll(keepFiles, oldSrcFiles));
     deleteDirFiles(binPath);
     writeFiles(srcPath, r.genFiles);

--- a/grammars/silver/core/List.sv
+++ b/grammars/silver/core/List.sv
@@ -173,6 +173,29 @@ function filter
 }
 
 @{--
+ - Monadic (actually Applicative) version of filter
+ -
+ - @param f  The filter function
+ - @param lst  The input list to filter
+ - @return  Only those elements of 'lst' that 'f' returns true for, in the
+ -   same order as they appeared in 'lst'
+ -}
+function filterM
+Applicative m =>
+m<[a]> ::= f::(m<Boolean> ::= a) lst::[a]
+{
+  return
+    case lst of
+    | [] -> pure([])
+    | h :: t -> do {
+        cond::Boolean <- f(h);
+        rest::[a] <- filterM(f, t);
+        return if cond then h :: rest else rest;
+      }
+    end;
+}
+
+@{--
  - Partition a list in two
  -
  - @param f  Decision function

